src/lib/peers/peer-connection/PeerServer.ts

```diff
@@ -0,0 +1,121 @@
+    this.server.addService(PeerService, {
+      messagePeer: this.messagePeer.bind(this),
+    });
+
+    // Create the server credentials. SSL only if ca cert exists
+    const pkiInfo = this.keyManager.PKIInfo;
+
+    if (pkiInfo.caCert && pkiInfo.cert && pkiInfo.key) {
+      this.credentials = grpc.ServerCredentials.createSsl(
+        pkiInfo.caCert,
+        [
+          {
+            private_key: pkiInfo.key,
+            cert_chain: pkiInfo.cert,
+          },
+        ],
+        true,
+      );
+    } else {
+      this.credentials = grpc.ServerCredentials.createInsecure();
+    }
+
+    const port = process.env.PK_PORT ?? this.peerManager.peerInfo.connectedAddress?.port ?? 0;
+    this.server.bindAsync(`0.0.0.0:${port}`, this.credentials, async (err, boundPort) => {
+      if (err) {
+        throw err;
+      } else {
+        const address = new Address('0.0.0.0', boundPort);
+        this.server.start();
+        this.peerManager.peerInfo.connectedAddress = address;
+        console.log(`Peer Server running on: ${address}`);
+        this.started = true;
+        this.turnServer = new TurnServer(this.peerManager);
+      }
+    });
+  }
+
+  private async messagePeer(call, callback) {
+    const peerRequest: PeerMessage = call.request;
+
+    const { publickey: publickey, type, submessage } = peerRequest.toObject();
+
+    // if we don't know publicKey, end connection
+    if (!this.peerManager.hasPeer(publickey)) {
+      throw Error('unknown public key');
+    }
+
+    // verify and decrypt request
+    const verifiedMessage = await this.keyManager.verifyData(Buffer.from(submessage), Buffer.from(publickey));
+    const decryptedMessage = await this.keyManager.decryptData(verifiedMessage);
+    const request = stringToProtobuf(decryptedMessage.toString());
+
+    let response: Uint8Array;
+    switch (type) {
+      case SubServiceType.PING_PEER:
+        response = await this.handlePing(request);
+        break;
+      case SubServiceType.GIT:
+        response = await this.handleGitRequest(request, publickey);
+        break;
+      case SubServiceType.NAT_TRAVERSAL:
+        response = await this.handleNatRequest(request);
+        break;
+      default:
+        throw Error('peer message type not identified');
+    }
+
+    // encrypt and sign response
+    const encryptedResponse = await this.keyManager.encryptData(
+      Buffer.from(protobufToString(response)),
+      Buffer.from(publickey),
+    );
+    const signedResponse = await this.keyManager.signData(encryptedResponse);
+    const subMessage = signedResponse.toString();
+
+    // composes peer message
+    const peerResponse = new PeerMessage();
+    peerResponse.setPublickey(this.peerManager.peerInfo.publicKey);
+    peerResponse.setType(type);
+    peerResponse.setSubmessage(subMessage);
+
+    // return peer response
+    callback(null, peerResponse);
+  }
+
+  private async handlePing(request: Uint8Array): Promise<Uint8Array> {
```

Comment:
This looks like a direct one to one ping. Not the sonar ping we were talking about. We should clarify the naming of these concepts.

Reply:
yeah you're right, this is more involved in checking if the peer node is alive which could probably be done other ways with gRPC. As opposed to our sonar ping which is more like scanning the multicast broadcast address for broadcasting peers right?